src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+    ///  foreach(KeyValuePair&lt;T, U&gt;> item in dictionary)
+    ///  { ... }
+    /// </summary>
+    /// <remarks>
+    ///  DeferredDictionary doesn't record anything but the file position of the dictionary in the JSON initially.
+    ///  If you foreach over the KeyValuePairs, it constructs one reader and loads the items as you read them.
+    ///  If you ask for the Keys, Values, or Dictionary[key], it must build a map from each key to the file position for that value,
+    ///  and must seek in the file for each read.
+    ///  
+    ///  Items are expensive to iterate each time; they are not kept in memory. Copy the values to a List or array to keep them.
+    /// </remarks>
+    /// <typeparam name="T">Type of items in Dictionary</typeparam>
+    public class DeferredDictionary<T> : IDictionary<string, T>
+    {
+        private JsonSerializer _jsonSerializer;
+        private Func<Stream> _streamProvider;
+        private long _start;
+
+        private Stream _stream;
+        private Dictionary<int, long> _itemPositions;
+
+        public DeferredDictionary(JsonSerializer jsonSerializer, JsonPositionedTextReader reader, bool buildPositionsNow = false)
+        {
+            _jsonSerializer = jsonSerializer;
+            _streamProvider = reader.StreamProvider;
+            _start = reader.TokenPosition;
+
+            // We have the JsonTextReader, which must scan to after the collection to resume building the outer object
+            // We may as well make the map of element positions.
+            if (buildPositionsNow)
+            {
+                BuildPositions(reader, 0);
+            }
+            else
+            {
+                reader.Skip();
+            }
+        }
+
+        private void EnsurePositionsBuilt()
+        {
+            if (_itemPositions == null) BuildPositions();
+        }
+
+        private void BuildPositions()
+        {
+            using (Stream stream = _streamProvider())
+            using (JsonPositionedTextReader reader = new JsonPositionedTextReader(() => stream))
+            {
+                stream.Seek(_start, SeekOrigin.Begin);
+                reader.Read();
+
+                BuildPositions(reader, _start);
+            }
+        }
+
+        private void BuildPositions(JsonPositionedTextReader reader, long currentOffset)
+        {
+            Dictionary<int, long> result = new Dictionary<int, long>();
+            while (true)
+            {
+                // Find the position just before the PropertyName (we need to read it back later to confirm the string matches)
+                long keyPosition = currentOffset + reader.TokenPosition + 1;
+
+                reader.Read();
+                if (reader.TokenType == JsonToken.EndObject) break;
+
+                if (reader.TokenType != JsonToken.PropertyName) throw new InvalidDataException($"@({reader.LineNumber}, {reader.LinePosition}): Expected property name, found {reader.TokenType} \"{reader.Value}\".");
```

Comment:
>if [](start = 16, length = 2)

this is a rather long line. more NITS for you to consider. recall my NIT around always including braces even for single lines (as in the line previous). i will stop remarking on this now.